core/src/sumeragi/network_topology.rs

```diff
@@ -139,13 +139,18 @@ impl Topology {
     }
 
     /// Add or remove peers from the topology.
-    pub fn update_peer_list(&mut self, new_peers: UniqueVec<PeerId>) {
-        self.ordered_peers = new_peers
-    }
-
-    /// Rotate peers after each failed attempt to create a block.
-    pub fn rotate_all(&mut self) {
-        self.rotate_all_n(1);
+    pub fn update_peer_list(&mut self, other_new_peers: UniqueVec<PeerId>) {
+        let mut new_peers: Vec<_> = other_new_peers.into();
+        let mut ordered_peers: Vec<_> = self.ordered_peers.clone().into();
+        ordered_peers.retain(|p| match new_peers.iter().position(|p2| p2 == p) {
+            Some(index) => {
+                new_peers.remove(index);
+                true
+            }
+            None => false,
+        });
+        ordered_peers.append(&mut new_peers);
+        self.modify_peers_directly(|peers| *peers = ordered_peers);
```

Comment:
We can automatically dedup with extend.

```suggestion
self.modify_peers_directly(|peers| peers.retain(|peer| new_peers.contains(peer)));
self.ordered_peers.extend(new_peers);
```